### PR TITLE
Multi line chained operator syntax

### LINKED ISSUE #7904: Chained operator block syntax
Spec: https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#chained-operator-code-blocks

*Example:*
```
nums = 1..100
    . each random
    . sort
    . take 100
```

*Should be equivalent to:*
```nums = (((1..100) . each random) . sort) . take 100```

This is implemented in the parser but not yet supported in translation. I think it could be desugared to applications in the IR?

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #7904. 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides. 
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
@kazcw, is there a documentation or tests of the chained operator syntax other than [this single paragraph](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#chained-operator-code-blocks)?

What happens when the operator isn't `.`, but something else? Is it an error (just like in my PR)?

**Comment by somebody1234** [comment]
there seem to be more tests here:
https://github.com/enso-org/enso/blob/5216b8c0cb7991f71e68cfb0e528f7ca7219ec3d/lib/rust/parser/debug/tests/parse.rs#L458-L481

the first case contains other operators:
```py
value = nums
    * each random
    + constant
```

the second case contains a nested `OperatorBlockApplication`:
```py
rect1
    . width = 7
    . center
        + x
```
note that as per the design, it should be semantically equivalent to `(((rect1) . width = 7) . center + x)`

but at the very least, it means that the parser supports both these cases - i guess it's not explicitly part of the design, however the design _does_ always say "an operator" rather than "the dot operator"

File: engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java [hunk 1]
```diff
@@ -1246,6 +1246,40 @@ public void testSkip() throws Exception {
...
+    nums v fm ff n = v
+        . map fm
+        . filter ff
+        . take n
+    """);
+  }
+
+  @Test
+  public void testBlockSyntaxOperators() throws Exception {
+    equivalenceTest("""
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java:1262]: Thanks for [your help](https://github.com/enso-org/enso/pull/8415#issuecomment-1831264370) @somebody1234 - here is the same test in `runtime-parser` project.

However I have a question: what shall happen with operator priorities? If I use `+` on the first operator line and `*` on the second operator line, shall the order be kept or should `*` take precedence? CCing @kazcw

Comment (somebody1234) [reply engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java:1262]: i edited my original comment - the spec seems to say it should be line by line - see the last line of the spec

Comment (somebody1234) [reply engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java:1262]: spec excerpt, relevant parts highlighted:

image

File: engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java [hunk 2]
```diff
@@ -1246,6 +1246,40 @@ public void testSkip() throws Exception {
...
+
+  @Test
+  public void testBlockSyntaxOperators2() throws Exception {
+    equivalenceTest("""
+    value = (nums + each random) * constant
+    """, """
+    value = nums
+        + each random
+        * constant
+    """);
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java:1279]: According to @somebody1234's [comment](https://github.com/enso-org/enso/pull/8415/files#r1408921830) this is the desired behavior.

File: engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java:1262]: Thank you. The expression is _"built line-by-line"_ as the [next test](https://github.com/enso-org/enso/pull/8415/files#r1411551881) shows.

**Comment by JaroslavTulach** [comment]
@somebody1234, @kazcw I am not sure what to think about

> the second case contains a nested `OperatorBlockApplication`:
> 
> ```python
> rect1
>     . width = 7
>     . center
>         + x
> ```
> 
> note that as per the design, it should be semantically equivalent to `(((rect1) . width = 7) . center + x)`

I believe it is not a valid syntax: what should it mean `rect1.width = 7`? Assignment to a field `width`? There are no assignments to fields in Enso.

**Comment by somebody1234** [comment]
@JaroslavTulach yeah... i'd assume that one is just a parser test. `((rect1) . width = 7).center` makes even less sense, i think

**Comment by JaroslavTulach** [comment]
The other inquiry I have is with nested blocks

> ```python
> rect1
>     . width
>     . center
>         + x
> ```

what's the difference between the above and no-nesting block?

```ruby
rect1
     . width
     . center
     + x
```

the latter one is clearly `((rect1 . width) . center) + x` according to the [short spec](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#chained-operator-code-blocks), but the first one? Does it have the same meaning (why do we have double syntax meaning the same)? Or should it first evaluate the nested block - e.g. `(center + x)`.

**Review by Akirathan — APPROVED** [review]

**Review by hubertp — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
The behavior probably isn't fully correct, but the specification is so fuzzy that it's not clear what is correct anyway. Let's integrate at current state and solve the rest as bugs.

**Pull request merged**